### PR TITLE
refactor: introduce SecretValue to represent secrets

### PR DESCRIPTION
**Pull request by eladb**
The class `cdk.SecretValue` is a token that represents a value that should
be treated as a secret. It can be created through one of the factory methods:
`SecretValue.secretsManager(id)` or `SecretValue.plainText(text)`.

Constructs that reference secrets have been converted to use this type
instead of plain strings.

The `secretsmanager.Secret` construct now has `secretValue` and 
`secretJsonValue` which return `SecretValue` objects with a dynamic reference.

The `DynamicReference` class is now a `Token`.

BREAKING CHANGE: The `secretsmanager.SecretString` class has been removed in favor of `cdk.SecretValue.secretsManager(id[, options])`

The following prop types have been changed from `string` to `cdk.SecretValue`: `codepipeline-actions.AlexaSkillDeployAction.clientSecret`, `codepipeline-actions.AlexaSkillDeployAction.refreshToken`, `codepipeline-actions.GitHubSourceAction.oauthToken`, `iam.User.password`

`secretsmanager.Secret.stringValue` and `jsonFieldValue` have been removed. Use `secretsmanage.Secret.secretValue` and `secretJsonValue` instead.

`secretsmanager.Secret.secretString` have been removed. Use `cdk.SecretValue.secretsManager()` or `secretsmanager.Secret.import(..).secretValue`.

The class `cdk.Secret` has been removed. Use `cdk.SecretValue` instead.

The class `cdk.DynamicReference` is no longer a construct, and it's
constructor signature was changed.

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/cdk/lib/secret-value.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  public static plainText(secret: string): SecretValue {
+    return new SecretValue(secret);
+  }
+
+  /**
+   * Creates a `SecretValue` with a value which is dynamically loaded from AWS Secrets Manager.
+   * @param secretId The ID or ARN of the secret
+   * @param options Options
+   */
+  public static secretsManager(secretId: string, options: SecretsManagerSecretOptions = { }): SecretValue {
```
Comment (rix0rrr) [inline packages/@aws-cdk/cdk/lib/secret-value.ts]: If you're going to go the way of adding constructor functions here, then you should also add one for SSMPS.

**Comment by jogold** [comment]
Hey @eladb, shouldn't this also be converted?
https://github.com/awslabs/aws-cdk/blob/997dbcc33fa6802ae807db07a392a7382df7dc7c/packages/%40aws-cdk/aws-rds/lib/props.ts#L77

File: packages/@aws-cdk/cdk/lib/secret-value.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/cdk/lib/secret-value.ts]: Good point.

Comment (rix0rrr) [reply packages/@aws-cdk/cdk/lib/secret-value.ts]: And it should kinda/sorta also be possible to use a NoEcho CfnParameter.

We shouldn't encourage it but it is something people can do today, and SecretsManager *does* cost money per secret per month.

Comment (eladb) [reply packages/@aws-cdk/cdk/lib/secret-value.ts]: Yap adding that too. Enforcing NoEcho is a nice touch.

**Pull request merged**